pkg/archive/filter.go

```diff
@@ -0,0 +1,78 @@
+package archive
+
+import (
+	"bufio"
+	"io"
+	"os/exec"
+	"sync"
+
+	"github.com/containers/storage/pkg/pools"
+)
+
+var (
+	procFilterLock sync.Mutex
+	filterPath     map[string]string
+)
+
+type filterReader struct {
+	reader *io.PipeReader
+	runErr error
+}
+
+func (f *filterReader) Read(p []byte) (n int, err error) {
+	if f.runErr != nil {
+		return 0, f.runErr
+	}
+	return f.reader.Read(p)
+}
+
+func (f *filterReader) Close() error {
+	err := f.reader.Close()
+	if err == nil {
+		err = f.runErr
+	}
+	return err
+}
+
+func getFilterPath(name string) string {
+	procFilterLock.Lock()
+	if filterPath == nil {
+		filterPath = make(map[string]string)
+	}
+	path, ok := filterPath[name]
+	procFilterLock.Unlock()
+
+	if ok {
+		return path
+	}
+
+	path, err := exec.LookPath(name)
+	if err != nil {
+		path = ""
+	}
+
+	procFilterLock.Lock()
+	filterPath[name] = path
+	procFilterLock.Unlock()
+	return path
+}
+
+func tryProcFilter(args []string, pool *pools.BufioReaderPool, input *bufio.Reader) (io.ReadCloser, bool) {
+	path := getFilterPath(args[0])
+	if path == "" {
+		return nil, false
+	}
+
+	r, w := io.Pipe()
+	cmd := exec.Command(path, args[1:]...)
+	cmd.Stdin = input
+	cmd.Stdout = w
+
+	f := &filterReader{reader: r}
+	go func() {
+		defer w.Close()
+		f.runErr = cmd.Run()
```

Comment:
- This is a data race vs the consumer goroutine
- I don’t immediately see that this `filterReader` is necessary: Isn’t it enough to do

```go
var err = errors.New("internal error: should never be seen")
defer w.CloseWithError(err) // CloseWithErr(nil) == Close()
err = cmd.Run()
```

Reply:
under what conditions will `err` ever be `errors.New("internal error: should never be seen")`?

Reply:
Panic before `Run` returns.

Reply:
… and it’s a belt-and-suspenders thing: this makes sure to avoid a hypothetical `CloseWithError(nil)` on the `panic` path — whether or not `panic` is relevant, the point is that the claim “the pipe is successfully closed _only_ on success” can be easily verified.